cmd/posix.go

```diff
@@ -541,6 +541,99 @@ func (s *posix) ReadFile(volume string, path string, offset int64, buf []byte) (
 	return int64(m), err
 }
 
+func (s *posix) PrepareFile(volume, path string, fileSize int64) (err error) {
+	defer func() {
+		if err == syscall.EIO {
+			atomic.AddInt32(&s.ioErrCount, 1)
+		}
+	}()
+
+	if s.ioErrCount > maxAllowedIOError {
+		return errFaultyDisk
+	}
+
+	// Validate if disk is free.
+	if err = s.checkDiskFree(); err != nil {
+		return err
+	}
+
+	volumeDir, err := s.getVolDir(volume)
+	if err != nil {
+		return err
+	}
+	// Stat a volume entry.
+	_, err = os.Stat(preparePath(volumeDir))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return errVolumeNotFound
+		}
+		return err
+	}
+
+	filePath := pathJoin(volumeDir, path)
+	if err = checkPathLength(filePath); err != nil {
+		return err
+	}
+
+	// Verify if the file already exists and is not of regular type.
+	var st os.FileInfo
+	if st, err = os.Stat(preparePath(filePath)); err == nil {
+		if !st.Mode().IsRegular() {
+			return errIsNotRegular
+		}
+	}
+
+	// Create top level directories if they don't exist.
+	// with mode 0777 mkdir honors system umask.
+	if err = mkdirAll(filepath.Dir(filePath), 0777); err != nil {
+		// File path cannot be verified since one of the parents is a file.
+		if isSysErrNotDir(err) {
+			return errFileAccessDenied
+		} else if isSysErrPathNotFound(err) {
+			// Add specific case for windows.
+			return errFileAccessDenied
+		}
+		return err
+	}
+
+	w, err := os.OpenFile(preparePath(filePath), os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
+	if err != nil {
+		// File path cannot be verified since one of the parents is a file.
+		if isSysErrNotDir(err) {
+			return errFileAccessDenied
+		}
+		return err
+	}
+
+	// Close upon return.
+	defer w.Close()
+
+	if fileSize <= 0 {
```

Comment:
> Validate this earlier before creating a file and its parents. Otherwise this would unnecessarily leaves stale information.

yes, your are right, my original idea was about "preparing" a file so we do return nil even if fallocate fails because the file was already created at this point. But still like this it is more complicated. I'll fix it.